Two Step Section: Add a new setting (under a feature flag) to enforce security keys

xknown opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

## Proposed Changes

* Adds a new setting to enforce the use of security keys (passkeys) for a given WordPress.com account. This is currently hidden under a feature flag while the AR flows are improved.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Add a passkey if you don't have one in `/me/security/two-step/`
* Make sure you are able to toggle the new setting.

If the setting is enabled, the `re-auth` dialog now only shows the ability to use a passkey.
Before
<img width="401" alt="Screenshot 2024-01-25 at 16 16 27" src="https://github.com/Automattic/wp-calypso/assets/1081870/b9ec662c-fe20-45a8-b350-9788e5dc6d29">
After
<img width="400" alt="Screenshot 2024-01-25 at 16 18 50" src="https://github.com/Automattic/wp-calypso/assets/1081870/00fc0652-427b-41f9-9253-4b1cdfe9c458">



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

xknown committed:
Two Step Section: Add a new setting to enforce security keys

Add a new setting to enforce the use of security keys (passkeys) for a
given WordPress.com account.

xknown committed:
Remove unintended changes

xknown committed:
Hide the new setting behind a feature flag.

xknown committed:
Only use the flag for updating the setting, not for the re-auth dialog

tyxla commented on client/me/reauth-required/index.jsx:
```diff
@@ -170,7 +171,10 @@ class ReauthRequired extends Component {
 	}
 
-	renderVerificationForm() {
+	renderVerificationForm( enhancedSecurity ) {
```
We could just use the props and that way we won't need an argument:

```suggestion
	renderVerificationForm() {
		const enhancedSecurity = this.props.twoStepAuthorization.data?.two_step_enhanced_security;
```

tyxla commented on client/me/reauth-required/index.jsx:
```diff
@@ -245,30 +275,33 @@ class ReauthRequired extends Component {
 					/>
 				) : (
-					this.renderVerificationForm()
+					this.renderVerificationForm( enhancedSecurity )
```
If you agree on what I suggested above on retrieving this from props, let's remove this arg.

```suggestion
					this.renderVerificationForm()
```

tyxla commented on client/me/reauth-required/index.jsx:
```diff
@@ -232,7 +236,33 @@ class ReauthRequired extends Component {
+
+		return (
+			<Card compact>
+				<CardHeading tagName="h2">{ translate( 'Two Factor Auth Required' ) }</CardHeading>
```
I think we never use "auth" in front of users as it can be too techy for some. Perhaps we should alter this to "Authentication" to match the remaining strings.

xknown replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/85234/commits/7245669d4469ff5eba53b5cf2fedc3efda2984a6

tyxla commented on client/me/security-2fa-enhanced-security-setting/index.jsx:
```diff
@@ -0,0 +1,67 @@
+		<div className="security-2fa-enhanced-security-setting">
+			<SectionHeader label={ translate( 'Settings' ) }></SectionHeader>
+			<Card>
+				<form onChange={ onFormChange }>
```
Perhaps we could directly use `saveSettings` here?
```suggestion
				<form onChange={ saveSettings }>
```

tyxla commented on client/me/security-2fa-enhanced-security-setting/index.jsx:
```diff
@@ -0,0 +1,67 @@
+	const toggleSetting = ( settingValue ) => {
+		setSetting( 'two_step_enhanced_security', settingValue );
+	};
+	const onFormChange = () => {
+		saveSettings();
+	};
```
See below, but if we decide to use `saveSettings` in `form's `onChange`, we could completely remove this:

```suggestion
```

tyxla commented on client/me/security-2fa-enhanced-security-setting/index.jsx:
```diff
@@ -0,0 +1,67 @@
+	);
+};
+
+export default connect(
```
For functional components like this one it's recommended that we use `useSelector` and `useDispatch()` instead of `connect`.

xknown replied:
Done in https://github.com/Automattic/wp-calypso/pull/85234/commits/9f9d2783a4d9a76ab1bcc0fb4af092f45b76ecb6

tyxla commented on client/me/security-2fa-enhanced-security-setting/index.jsx:
```diff
@@ -0,0 +1,67 @@
+		setUserSetting,
+		saveUserSettings,
+	}
+)( localize( Security2faEnhancedSecuritySetting ) );
```
Same on using the `localize` high-order component - you could just use the `useTranslate()` hook in functional components.

tyxla commented on client/me/security-2fa-enhanced-security-setting/index.jsx:
```diff
@@ -0,0 +1,67 @@
+	}
+	return (
+		<div className="security-2fa-enhanced-security-setting">
+			<SectionHeader label={ translate( 'Settings' ) }></SectionHeader>
```
I think "Settings" is OK, but I'm just curious if there's a more specific heading we could use here. No preferences, personally, just wondering.

tyxla commented on client/me/security-2fa-enhanced-security-setting/index.jsx:
```diff
@@ -0,0 +1,67 @@
+import getUserSettings from 'calypso/state/selectors/get-user-settings';
+import { saveUserSettings, setUserSetting } from 'calypso/state/user-settings/actions';
+import { isFetchingUserSettings } from 'calypso/state/user-settings/selectors';
+import { recordGoogleEvent } from '../../state/analytics/actions';
```
Absolute paths are preferred:

```suggestion
import { recordGoogleEvent } from 'calypso/state/analytics/actions';
```

tyxla commented on config/wpcalypso.json:
```diff
@@ -179,6 +179,7 @@
 		"themes/text-search-lots": true,
 		"themes/assembler-first": true,
 		"themes/tiers": false,
+		"two-factor/enhanced-security": true,
```
Are we intentionally not enabling this on staging where more a12s could test it?

xknown replied:
Just an oversight, thanks for noticing it! :) - https://github.com/Automattic/wp-calypso/pull/85234/commits/8a964d79bc4b01e71b4a0647a9417b3fb0789157

tyxla commented on client/me/two-step/index.jsx:
```diff
@@ -89,9 +90,21 @@ class TwoStep extends Component {
+	renderEnhancedSecuritySetting = () => {
+		if (
+			! isEnabled( 'two-factor/enhanced-security' ) ||
+			this.props.isFetchingUserSettings ||
```
Do we mean to render the component if `isFetchingUserSettings` is true though? I see we're disabling the toggle field when fetching user settings, so we should be good to render it in that case?

xknown replied:
I was just following the other `render*` callbacks which don't render when `isFetchingUserSettings` is `true`. Is there a preferred option between not rendering vs disabled the setting?

tyxla replied:
That's totally fine to me, but then we don't need to disable the toggle in that case because a disabled toggle will never render anyway.

xknown replied:
Sounds good, removed it in 02c166fc41

tyxla reviewed (commented):
This is some great work 👍 

Thanks for prompting me to add a security key on the other laptop 😅 

I've left a few suggestions on the code style, but I'd say this is mostly looking good!

Let me know what you think.

jsnajdr commented on client/me/reauth-required/index.jsx:
```diff
@@ -245,30 +275,33 @@ class ReauthRequired extends Component {
 		const hasSmsRecoveryNumber =
 			!! this.props?.twoStepAuthorization?.data?.two_step_sms_last_four?.length;
 
+		const currentAuthType = enhancedSecurity ? 'webauthn' : twoFactorAuthType;
```
It would be better if `twoFactorAuthType` always had the right value. That can be achieved by:
1. Initializing the `twoFactorAuthType` state correctly in the component constructor. Now the initial value is hardcoded to `authenticator`, it should be `enhanced ? 'webauthn' : 'authenticator'`.
2. Not letting `TwoFactorActions` to switch it away from `webauth`. That probably already works.

jsnajdr commented on client/me/reauth-required/index.jsx:
```diff
@@ -245,30 +275,33 @@ class ReauthRequired extends Component {
 					isAuthenticatorSupported={ method !== 'sms' }
 					isSmsAllowed={ shouldEnableSmsButton }
 					isSecurityKeySupported={ isSecurityKeySupported }
+					requireSecurityKeyOnly={ enhancedSecurity }
```
Instead of this additional prop, can we change the values of the `is{Method}Supported` props to really say `false` if enhanced security is on? Then, for example, `isAuthenticationSupported` would mean simply "OTP is supported, period" and not the less clear "OTP is supported given that another prop is not `true`"

jsnajdr commented on client/me/security-2fa-enhanced-security-setting/index.jsx:
```diff
@@ -0,0 +1,67 @@
+			<Card>
+				<form onChange={ onFormChange }>
+					<FormFieldset>
+						<FormLabel>{ translate( 'Enhanced account security' ) }</FormLabel>
```
If I understand it correctly, enhanced security is a shortcut to disable all two-factor methods except webauthn, even though I have them set up? And when I disable the settings, I'll have all the old methods available again -- the same OTP key, the same SMS number...

I assume this will be force-set to true for all a12s, and probably disabled when they leave.

xknown replied:
That's correct.

And yes, we'll enforce it on the backend for a12s.

jsnajdr commented on client/me/security-2fa-enhanced-security-setting/index.jsx:
```diff
@@ -0,0 +1,67 @@
+		setUserSetting,
+		saveUserSettings,
+	}
+)( localize( Security2faEnhancedSecuritySetting ) );
```
This could have been a "modern" functional component that uses `useTranslate` and `useSelector` hooks instead of the HOCs, but nevermind, it's fine.

xknown replied:
Done in https://github.com/Automattic/wp-calypso/pull/85234/commits/9f9d2783a4d9a76ab1bcc0fb4af092f45b76ecb6

jsnajdr reviewed (approved):
Looks good! I have some suggestions about how to simplify and reorganize the code, but none are blockers, the code works well as is.

xknown committed:
Apply suggestions from code review

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

xknown committed:
Apply suggestions from code review

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

xknown committed:
Use a better card title when we show the unsupported security key
message.

xknown committed:
Enable this setting also on staging.

xknown committed:
Use high order components instead.

xknown committed:
Remove `isFetchingUserSettings` on `Security2faEnhancedSecuritySetting`

We already hide this component when `isFetchingUserSettings` is `true`.
Let's remove the redundant code.

tyxla reviewed (approved):
Nothing further from me, this looks great! 

One last thing, we should enable this in production once we're ready for that - but I'd assume we want to test internally first. 

Thanks @xknown, excited to be adopting security keys more broadly!

xknown commented:
Thank you very much for the review, @tyxla and @jsnajdr!

xknown committed:
Reorganize the code to avoid adding unnecessary code

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/11095392

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @xknown for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/11095392).
